pilot/pkg/xds/endpoint_builder.go

```diff
@@ -191,6 +298,33 @@ func (b *EndpointBuilder) buildLocalityLbEndpointsFromShards(
 	return locEps
 }
 
+// TODO(lambdai): Handle ApplyTunnel error return value by filter out the failed endpoint.
+func (b *EndpointBuilder) ApplyTunnelSetting(llbOpts []*LocLbEndpointsAndOptions, tunnelType networking.TunnelType) []*LocLbEndpointsAndOptions {
+	for _, llb := range llbOpts {
+		for i, ep := range llb.llbEndpoints.LbEndpoints {
+			newEp, err := llb.tunnelMetadata[i].ApplyTunnel(ep, tunnelType)
+			if err != nil {
+				panic("not implemented yet on failing to apply tunnel")
```

Comment:
nit: would prefer to either not return an error here (since it never fails) or handle the error instead of panic. I know from looking at the code it will never panic but that may change